src/main/java/org/radarbase/appserver/service/FcmNotificationService.java

```diff
@@ -1,22 +1,311 @@
+        }
+    }
+
+    @SneakyThrows
+    @Transactional
+    public FcmNotificationDto addNotificationForced(FcmNotificationDto notificationDto, String fcmToken, String subjectId, String projectId) {
+        Optional<Project> project = this.projectRepository.findByProjectId(projectId);
+        Project newProject;
+        if(project.isEmpty()) {
+            Project project1 = new Project();
+            newProject = this.projectRepository.save(project1.setProjectId(projectId));
+        } else {
+            newProject = project.get();
+        }
+
+        Optional<User> user = this.userRepository.findBySubjectIdAndProjectId(subjectId, newProject.getId());
+        AtomicReference<User> newUser = new AtomicReference<>();
+
+        user.ifPresentOrElse(newUser::set, () -> newUser.set(this.userRepository.save(
+                new User().setEnrolmentDate(Instant.now())
+                        .setFcmToken(fcmToken)
+                        .setTimezone(0)
+                        .setProject(newProject)
+                        .setSubjectId(subjectId))
+                .setUserMetrics(new UserMetrics(LocalDateTime.now().toInstant(ZoneOffset.UTC), null)
+                )));
+
+        if(!notificationRepository.existsByUserIdAndSourceIdAndScheduledTimeAndTitleAndBodyAndTypeAndTtlSeconds(
+                newUser.get().getId(),
+                notificationDto.getSourceId(),
+                notificationDto.getScheduledTime().toInstant(ZoneOffset.UTC),
+                notificationDto.getTitle(),
+                notificationDto.getBody(),
+                notificationDto.getType(),
+                notificationDto.getTtlSeconds()
+        )) {
+            Notification notification = ConverterFactory.getNotificationConverter()
+                    .dtoToEntity(notificationDto).setUser(newUser.get());
+            notification = this.notificationRepository.save(notification);
+
+            this.schedulerService.scheduleNotification(notification);
+
+            return ConverterFactory.getNotificationConverter().entityToDto(notification);
+        } else {
+            throw new AlreadyExistsException("The Notification Already exists. Please Use update endpoint", notificationDto);
+        }
+    }
+
+    @Transactional
+    public FcmNotificationDto updateNotification(FcmNotificationDto notificationDto, String subjectId, String projectId) {
+        Optional<Project> project = this.projectRepository.findByProjectId(projectId);
+        if(project.isEmpty()) {
+            throw new NotFoundException("Project Id does not exist. Please create a project with the ID first");
+        }
+
+        Optional<User> user = this.userRepository.findBySubjectIdAndProjectId(subjectId, project.get().getId());
+        if(user.isEmpty()) {
+            throw new NotFoundException("The supplied subject ID is invalid. No user found. Please Create a User First.");
+        }
+        if(notificationDto.getId() == null) {
```

Comment:
Do this check before querying for project and subject

Reply:
Good point. Thanks